File: client/src/main/java/io/xol/chunkstories/input/lwjgl3/Lwjgl3MouseButton.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.input.lwjgl3;

import static org.lwjgl.glfw.GLFW.GLFW_PRESS;
import static org.lwjgl.glfw.GLFW.glfwGetMouseButton;

import io.xol.chunkstories.api.client.ClientInterface;
import io.xol.chunkstories.api.input.Input;
import io.xol.chunkstories.api.input.Mouse.MouseButton;
import io.xol.chunkstories.client.Client;

public class Lwjgl3MouseButton implements MouseButton, LWJGLPollable
{
	private final Lwjgl3Mouse mouse;
	private final String name;
	private final int button;
	
	private boolean isDown = false;
	
	public Lwjgl3MouseButton(Lwjgl3Mouse mouse, String name, int button)
	{
		this.mouse = mouse;
		
		this.name = name;
		this.button = button;
	}
	
	@Override
	public String getName()
	{
		return name;
	}

	@Override
	public boolean isPressed()
	{
		return isDown;
	}
	
	public long getHash()
	{
		return button;
	}

	@Override
	public void updateStatus()
	{
		isDown = glfwGetMouseButton(mouse.im.gameWindow.glfwWindowHandle, button) == GLFW_PRESS;
	}

	@Override
	public ClientInterface getClient() {
		return Client.getInstance();
	}

	@Override
	public Lwjgl3Mouse getMouse() {
		return mouse;
	}
	
	@Override
	public boolean equals(Object o)
	{
		if(o == null)
			return false;
		else if(o instanceof Input) {
			return ((Input)o).getName().equals(getName());
		}
		else if(o instanceof String) {
			return ((String)o).equals(this.getName());
		}
		return false;
	}
	
	@Override
	public int hashCode()
	{
		return getName().hashCode();
	}

}

File: client/src/main/java/xyz/chunkstories/input/lwjgl3/Lwjgl3MouseButton.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.input.lwjgl3

import org.lwjgl.glfw.GLFW.GLFW_PRESS
import org.lwjgl.glfw.GLFW.glfwGetMouseButton

import xyz.chunkstories.api.client.Client
import xyz.chunkstories.api.input.Input
import xyz.chunkstories.api.input.Mouse.MouseButton
import xyz.chunkstories.input.AbstractInput
import xyz.chunkstories.input.Pollable

class Lwjgl3MouseButton(override val mouse: Lwjgl3Mouse, name: String, private val button: Int) : AbstractInput(mouse.inputsManager, name), MouseButton, Pollable {

    var isDown = false

    override val isPressed: Boolean
        get() {
            val ingameClient = mouse.inputsManager.gameWindow.client.ingame
            return if (ingameClient != null) isDown && ingameClient.player.hasFocus() else isDown
        }

    override fun updateStatus() {
        isDown = glfwGetMouseButton(mouse.inputsManager.gameWindow.glfwWindowHandle, button) == GLFW_PRESS
    }

    override val client: Client
        get() = mouse.inputsManager.gameWindow.client

    override fun equals(o: Any?) = when (o) {
        null -> false
        is Input -> o.name == name
        is String -> o == this.name
        else -> false
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }

}
